### ISSUE TITLE
For the TUDataset, can we separate the labels and features of node / edge?

### ISSUE THREAD

**Issue by jinhw1989** [rusty1s/pytorch_geometric#531]
## ❓ Questions & Help

I see the implementations of TUDataset
https://github.com/rusty1s/pytorch_geometric/blob/55c2e7da1840433e3331069e80deba481e4d00a4/torch_geometric/read/tu.py#L27-L38

where the node labels are concatenated to node features.
A similar process of edges can be seen as well.

Can we separate the labels and features?

In some cases, I may learn the node representations from just one of them, either labels or features, rather than both of them. 
For example, a recent paper of label-aware GCN https://arxiv.org/abs/1907.04707, use the labels of neighborhood edges.

Basically, let the user decide what kind of features send to GNN.

**Comment by rusty1s** [comment]
There is nothing that prevent you from separating the labels and features. For example, the `TUDataset` provides the attribute `use_node_attr` to only use node labels (note that it is set to `False` by default). If you only want to use node attributes, you can use the `num_node_labels` and `num_node_attributes` properties of the `TUDataset`, and, *e.g.*, select the right features using custom transforms:
```python
dataset = TUDataset(..., use_node_attr=True)

class MyCustomTransform(class):
    def __call__(self, data):
        data.x = data.x[:, :dataset.num_node_attributes]
        return data

dataset.transform = MyCustomTransform()
```
I agree that this is not really elegant, but I guess this use-case is quite rare.

**Comment by jinhw1989** [comment]
Thanks. 
BTW, it is better to use a sparse tensor when the feature is built from labels only, just for saving memory.

**Comment by rusty1s** [comment]
While technically true, this introduces a lot of other problems. This works well, *e.g.*, for `GCN`, however for other operators (that need to aggregate those sparse features) this basically results in sparse-sparse matrix multiplications. This is *very* slow in contrast to simply using dense features, and will only support gradients w.r.t. to non-zero values.

In the end, it is not worth the hazzle for low-dimensional initial node features. The operators that can naturally support sparse node feature matrices will accept them though, so feel free to convert your matrices to sparse ones if you feel that memory usage is a problem.

**Comment by jinhw1989** [comment]
Really helpful. Thank you. 
Close it for now.
